### PR TITLE
Simplify dashboard template for summary view

### PR DESCRIPTION
**Pull request by ongspxm**
changes to shift to a single template system where content is dynamically loaded. wip.
here just for the travis test.

mostly a transfer of features, with a more vue-powered structure. major architectural changes, but most doesn't affect the output, here are some that do.

**proposed message**
```
There are 2 main templates, which are loaded separately to process the
information when it is required. this requires the same templates to be
generated multiple times for the data. This greatly increases the size of the
generated report uneccessarily, which ideally only contains data for the report.

Let's split up the report generation process and dashboard.

This PR supports the display of the commit history. Further work on the panel
for authorship information will be addressed in another PR.
```

## things to consider
- [x] scale of ramp slices & contribution bar

## changes to data results
- weekly commits calculated from daily commits

### result not displayed immediately, loaded from hash or user input
image
this single file template is capable of being used with different set of results data.

### changing display style to make it obvious when grouped by repos
image
currently when grouped by repo, it is hard to tell, but this manner of grouping them with a bolded title allows users to identify different repos.

### CODE REVIEW

**Comment by damithc** [comment]
1. Try left-aligning the repo name. That way the readers eye doesn't have to zig-zag across the page to locate a specific team/author.
2. Would be nice to number the repos too, but can be a separate PR.
3. I'm wondering if a shorter font (as opposed to a one that has thin and tall letters) can help optimize vertical space. May be all fonts take the same vertical space. Not sure myself.

**Comment by ongspxm** [comment]
image
short and fat font, with team name aligned left

**Comment by ongspxm** [comment]
**group by repo**
image

**everything together**
bolded repo for easier identification
image

**Comment by damithc** [comment]
UI looks good

**Comment by ongspxm** [comment]
ready for review @yong24s @eugenepeh @yamidark @arrrhghgh

**Comment by ongspxm** [comment]
eh... codacy is such a bugger leh... i do so long oni do halfway

**Comment by ongspxm** [comment]
plus got some like impossible to fix leh... `Generic Object Injection Sink (security/detect-object-injection)`

**Comment by eugenepeh** [comment]
> plus got some like impossible to fix leh... Generic Object Injection Sink (security/detect-object-injection)

Those impossible to fix is fine to leave it. Try to fix all the code quality first.

**Comment by ongspxm** [comment]
complexity leh? i ignore also ah? dun make sense to break up the functions.

**Comment by damithc** [comment]
> complexity leh? i ignore also ah? dun make sense to break up the functions.

Avoid Singlish in comments. We cannot assume all future readers of this comment will be familiar with Singlish.

**Comment by ongspxm** [comment]
roger that prof, edited

**Comment by ongspxm** [comment]
ignoring `security/detect-object-injection` and `code complexity` errors, @eugenepeh @AdityaA1998 @yong24s @yamidark ready for review

**Comment by ongspxm** [comment]
eh, you must start your python server in the `<date>`, in the folder should have `v2`. it should find the file.

**Comment by yamidark** [comment]
@ongspxm ok can see it now, I think I forgot to `./gradlew build` it earlier =O

File: docs/user_guide.md [hunk 1]
```diff
@@ -130,3 +130,13 @@ Below is the list of features in this panel:
...
 - The lines that are NOT written by the current author will be marked gray, but they will still be displayed to provide context for the user
 - Segments of codes that are not written by the current author is default to be collapsed. If the user wants to refer to it, he can open the segment.
+
+## User Dashboard V2 (beta)
+Works is in progress for the second version of the dashboard. Instructions to get it up and running:
+- start a webserver serving your generated report directory
+- `python2 -m SimpleHTTPServer`
+- go to `http://localhost:8000/v2` on your browser, or whatever port your server is hosted
+- in the data field, enter `http://localhost:8000` and click go (again, your own port num should be used)
+- and viola.
```
Comment (yamidark) [inline docs/user_guide.md]: Note your English 😅

**Comment by eugenepeh** [comment]
Nits in message

~Currently, t~`T`here are 2 main templates, ~and each is~`which are` loaded separately to process
the information when it is required. this requires the same templates to be
generated multiple times for ~the data.~ ?? and why is this a problem?

Let's split up the report generation process and dashboard. 

~Also, t~This ~current
template only~ `PR` supports the display of the commit history. `Further work on t`~T~he panel for
authorship information will be ~worked on~`addressed` in another PR.

File: frontend/src/static/js/api.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
+/* util funcs */
+function $(id){ return document.getElementById(id); }
```
Comment (yamidark) [inline frontend/src/static/js/api.js]: Do a line break for the this (like `enquery` below)

File: frontend/src/static/js/api.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        loadJSON(REPORT_DIR+"/summary.json", (repos) => {
+            var names = [];
+            for(var repo of repos){
+                var repoName = repo.organization+"_"+repo.repoName;
+                window.REPOS[repoName] = repo;
+                names.push(repoName);
+            }
+
+            if(callback){ callback(); }
+            for(var name of names){ api.loadCommits(name); }
```
Comment (yamidark) [inline frontend/src/static/js/api.js]: line break as well

File: frontend/src/static/js/api.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+                    displayName: commits.authorDisplayNameMap[author],
+                    weeklyCommits: commits.authorWeeklyIntervalContributions[author],
+                    dailyCommits: commits.authorDailyIntervalContributions[author],
+                    totalCommits: commits.authorFinalContributionMap[author],
+                };
+                obj.searchPath = REPOS[repo].organization+"/";
+                obj.searchPath += REPOS[repo].repoName+"/";
+                obj.searchPath += obj.displayName+"/";
+                obj.searchPath += author;
+                obj.searchPath = obj.searchPath.toLowerCase();
```
Comment (yamidark) [inline frontend/src/static/js/api.js]: Is there a need for the `obj.searchPath` on each line?
Can we just do `+` at each line?

File: frontend/src/static/js/main.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        },
+        addUsers(users) {
+            this.userUpdated = false;
+            this.loadedRepo += 1;
+            this.userUpdated = true;
+        },
+        getUsers() {
+            var full = [];
+            for(var repo in this.repos){
+                if(!this.repos[repo].users){ continue; }
```
Comment (yamidark) [inline frontend/src/static/js/main.js]: line break here as well

File: frontend/src/static/js/v_summary.js [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+    },
+    computed: {
+        sliceCount() { return this.filtered[0][0].commits.length; },
+        sliceWidth() { return 100/this.sliceCount; },
+        avgCommitSize() {
+            var totalCommits=0, totalCount=0;
+            for(var repo of this.filtered){
+                for(var user of repo){
+                    for(var slice of user.commits){
+                        if(slice.insertions===0){ continue; }
```
Comment (yamidark) [inline frontend/src/static/js/v_summary.js]: line break here as well

File: frontend/src/static/js/v_summary.js [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+                    }
+                }
+            }
+            return totalCommits/totalCount;
+        },
+        avgContributionSize() {
+            var totalLines=0, totalCount=0;
+            for(var repo of this.filtered){
+                for(var user of repo){
+                    if(user.totalCommits===0){ continue; }
```
Comment (yamidark) [inline frontend/src/static/js/v_summary.js]: line break here as well

File: frontend/src/static/js/v_summary.js [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+        getSliceLink(user, slice) {
+            var REPOS = window.REPOS;
+
+            return "http://github.com/" +
+              REPOS[user.repoId].organization + "/" +
+              REPOS[user.repoId].repoName + "/commits/" +
+              REPOS[user.repoId].branch + "?" +
+              "author=" + user.name + "&" +
+              "since=" + slice.fromDate + "&" +
+              "until=" + slice.toDate;
```
Comment (yamidark) [inline frontend/src/static/js/v_summary.js]: do the line breaks before the `+` operator:
`+ REPOS[user.repoId].organization + "/"`...

File: frontend/src/static/js/v_summary.js [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+        },
+        getContributionBars(totalContribution) {
+            var res = [];
+            var contributionLimit = (this.avgContributionSize*2);
+
+            var cnt = parseInt(totalContribution/contributionLimit);
+            for(var cntId=0; cntId<cnt; cntId++){ res.push(100); }
```
Comment (yamidark) [inline frontend/src/static/js/v_summary.js]: line break here as well

File: frontend/src/static/js/v_summary.js [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+
+            var last = (totalContribution%contributionLimit)/contributionLimit;
+            if(last!==0){ res.push(last*100); }
```
Comment (yamidark) [inline frontend/src/static/js/v_summary.js]: line break here as well

File: frontend/src/static/js/v_summary.js [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+            }
+
+            var minDate="", maxDate="";
+            for(var repo of this.filtered){
+                for(var user of repo){
+                    var commits = user.commits;
+                    var date1 = commits[0].fromDate;
+                    var date2 = commits[commits.length-1].fromDate;
+                    if(!minDate || minDate>date1){ minDate=date1; }
+                    if(!maxDate || maxDate<date2){ maxDate=date2; }
```
Comment (yamidark) [inline frontend/src/static/js/v_summary.js]: line breaks here as well

File: frontend/src/static/js/v_summary.js [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+                }
+            }
+
+            if(!this.filterSinceDate){ this.filterSinceDate=minDate; }
+            if(!this.filterUntilDate){ this.filterUntilDate=maxDate; }
```
Comment (yamidark) [inline frontend/src/static/js/v_summary.js]: line breaks here as well

File: frontend/src/static/js/v_summary.js [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+            var full = [];
+
+            for(var repo of this.repos){
+                var res = [];
+
+                // filtering
+                for(var user of repo.users){
+                    if(user.searchPath.search(this.filterSearch)>-1){
+                        this.getUserCommits(user);
+                        if(this.filterGroupWeek){ this.splitCommitsWeek(user); }
```
Comment (yamidark) [inline frontend/src/static/js/v_summary.js]: line break here as well

File: frontend/src/static/js/v_summary.js [hunk 14]
```diff
@@ -0,0 +1,278 @@
...
+                        res.push(user);
+                    }
+                }
+
+                if(res.length){ full.push(res); }
```
Comment (yamidark) [inline frontend/src/static/js/v_summary.js]: line break here as well

File: frontend/src/static/js/v_summary.js [hunk 15]
```diff
@@ -0,0 +1,278 @@
...
+        },
+        getUserCommits(user) {
+            user["commits"] = [];
+            var userFirst = user.dailyCommits[0];
+            var userLast = user.dailyCommits[user.dailyCommits.length-1];
+
+            var sinceDate = this.filterSinceDate;
+            var untilDate = this.filterUntilDate;
+            if(!sinceDate){ sinceDate = userFirst.fromDate; }
+            if(!untilDate){ untilDate = userLast.fromDate; }
```
Comment (yamidark) [inline frontend/src/static/js/v_summary.js]: line breaks here as well

File: frontend/src/static/js/v_summary.js [hunk 16]
```diff
@@ -0,0 +1,278 @@
...
+
+            if(this.filterGroupWeek){ sinceDate = dateRounding(sinceDate, 1); }
```
Comment (yamidark) [inline frontend/src/static/js/v_summary.js]: line break here as well

File: frontend/src/static/js/v_summary.js [hunk 17]
```diff
@@ -0,0 +1,278 @@
...
+                    insertions:0,
+                    deletions:0,
+                    fromDate:getDateStr(startMs + dayId*DAY),
+                    toDate:getDateStr(startMs + (dayId+1)*DAY)
+                });
+            }
+
+            for(var commit of user.dailyCommits){
+                if(commit.fromDate<sinceDate){ continue; }
+                if(commit.fromDate>untilDate){ break; }
```
Comment (yamidark) [inline frontend/src/static/js/v_summary.js]: line breaks here as well

File: frontend/src/static/js/v_summary.js [hunk 18]
```diff
@@ -0,0 +1,278 @@
...
+                user.commits.push(commit);
+            }
+
+            if(this.filterGroupWeek){ untilDate = dateRounding(untilDate); }
```
Comment (yamidark) [inline frontend/src/static/js/v_summary.js]: line break here as well

File: frontend/src/static/js/v_summary.js [hunk 19]
```diff
@@ -0,0 +1,278 @@
...
+            }
+
+            if(this.filterSortReverse){
+                for(var repo of full){ repo.reverse(); }
+            }
+
+            this.filtered = full;
+        }
+    },
+    created() { this.getFiltered(); }
```
Comment (yamidark) [inline frontend/src/static/js/v_summary.js]: line break here as well

**Comment by ongspxm** [comment]
@eugenepeh, @yamidark, @AdityaA1998, @yong24s ready for review

File: frontend/src/static/js/main.js [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+var app = new window.Vue({
+    el: "#app",
+    data: {
+        reportDirInput: "",
+        repos: {},
+        repoLength: 0,
+        loadedRepo: 0,
+        userUpdated: false
+    },
+    methods:{
```
Comment (yong24s) [inline frontend/src/static/js/main.js]: Inconsistent use of space before {.

File: frontend/src/static/js/v_summary.js [hunk 21]
```diff
@@ -0,0 +1,312 @@
+function comparator(fn){ return function(a, b){
+    var a1 = fn(a), b1=fn(b);
+    if(a1 === b1){ return 0; }
+    else if(a1 < b1){ return -1; }
+    return 1;
+};}
+
+/* dates funcs */
+const DAY = (1000*60*60*24);
```
Comment (yong24s) [inline frontend/src/static/js/v_summary.js]: Would ```MILLISECONDS_IN_A_DAY``` be more intuitive?

File: docs/user_guide.md [hunk 22]
```diff
@@ -131,3 +131,12 @@ Below is the list of features in this panel:
 - The user can click the title again to display the file content.
 - The lines that are NOT written by the current author will be marked gray, but they will still be displayed to provide context for the user
 - Segments of codes that are not written by the current author is default to be collapsed. If the user wants to refer to it, he can open the segment.
+
+## User Dashboard V2 (beta)
+Works is in progress for the second version of the dashboard. Instructions to get it up and running:
+- start a webserver serving your generated report directory
+- `python2 -m SimpleHTTPServer 8000` or `python3 -m http.server 8000`
+- go to `http://localhost:8000/v2` on your browser
+- in the data field, enter `http://localhost:8000`
```
Comment (eugenepeh) [inline docs/user_guide.md]: As mentioned earlier, avoid more complexity and dependency on additional software if necessary. 
Also, try to make it deploy-able on static web page - e.g. GitHub pages.
It is okay if Google chrome is the only one that's unable to display the page for now.

**Comment by ongspxm** [comment]
once green tick, then ready for review, i will do the support for local file in another PR, but probably will require some changes to zip the whole of the report into one zip file.

**Pull request merged**